socketshark/receiver.py

```diff
@@ -79,29 +83,35 @@ async def ping_handler(self):
             if wait:
                 wait.cancel()
             if not self._stop:
-                self.shark.log.exception('unhandled exception in ping handler')
+                self.shark.log.debug('redis ping handler cancelled')
         except Exception:
             self.shark.log.exception('unhandled exception in ping handler')
+            self._stop = True
         finally:
-            await self.stop()
+            if self._stop:
+                await self.stop()
 
     async def reader(self, once=False):
         self._stop = False
         try:
             ping_handler = asyncio.ensure_future(self.ping_handler())
-            result = await self._reader(once=once)
+            tasks = [
```

Comment:
Previously, calling `reader(once=True)` would guarantee that you read either 0 messages or 1. Now, it can be 0, 1, or 2. Is that right? Getting two messages with `once=True` seems misleading.

Reply:
running with `once=True` will actually read 0 or n messages. We repeat the loop in `_reader` until the `redis_receiver` queue is empty.

https://github.com/closeio/socketshark/blob/master/socketshark/receiver.py#L129-L130